py/visdom/server.py

```diff
@@ -428,6 +432,142 @@ def on_close(self):
             self.subs.pop(self.sid, None)
 
 
+# TODO condense some of the functionality between this class and the
+# original SocketHandler class
+class ClientSocketWrapper():
+    """
+    Wraps all of the socket actions in regular request handling, thus
+    allowing all of the same information to be sent via a polling interface
+    """
+    def __init__(self, app):
+        self.port = app.port
+        self.env_path = app.env_path
+        self.app = app
+        self.state = app.state
+        self.subs = app.subs
+        self.sources = app.sources
+        self.readonly = app.readonly
+        self.login_enabled = app.login_enabled
+        self.messages = []
+        self.last_read_time = time.time()
+        self.open()
+        try:
+            if not self.app.socket_wrap_monitor.is_running():
+                self.app.socket_wrap_monitor.start()
+        except AttributeError:
+            self.app.socket_wrap_monitor = tornado.ioloop.PeriodicCallback(
+                self.socket_wrap_monitor_thread, 15000
+            )
+            self.app.socket_wrap_monitor.start()
+
+    def socket_wrap_monitor_thread(self):
+        if len(self.subs) > 0:
+            for sub in list(self.subs.values()):
+                if time.time() - sub.last_read_time > MAX_SOCKET_WAIT:
+                    sub.close()
+        else:
+            self.app.socket_wrap_monitor.stop()
+
+    def broadcast_layouts(self, target_subs=None):
+        if target_subs is None:
+            target_subs = self.subs.values()
+        for sub in target_subs:
+            sub.write_message(json.dumps(
+                {'command': 'layout_update', 'data': self.app.layouts}
+            ))
+
+    def open(self):
+        if self.login_enabled and not self.current_user:
+            print("AUTH Failed in SocketHandler")
+            self.close()
+            return
+        self.sid = get_rand_id()
+        if self not in list(self.subs.values()):
+            self.eid = 'main'
+            self.subs[self.sid] = self
+        logging.info('Mocking new socket: {}'.format(self.sid))
+
+        self.write_message(
+            json.dumps({'command': 'register', 'data': self.sid,
+                        'readonly': self.readonly}))
+        self.broadcast_layouts([self])
+        broadcast_envs(self, [self])
+
+    def on_message(self, message):
+        logging.info('from web client: {}'.format(message))
+        msg = tornado.escape.json_decode(tornado.escape.to_basestring(message))
+
+        cmd = msg.get('cmd')
+
+        if self.readonly:
+            return
+
+        if cmd == 'close':
+            if 'data' in msg and 'eid' in msg:
+                logging.info('closing window {}'.format(msg['data']))
+                p_data = self.state[msg['eid']]['jsons'].pop(msg['data'], None)
+                event = {
+                    'event_type': 'Close',
```

Comment:
Does this need to be a lowercase `'close'` as well?

Reply:
In theory it should be - I hadn't observed it as the pane is already closed on the frontend someone is observing from 😅